### PR TITLE
[ECO-5117] Fix channel ATTACH/DETACH state checks

### LINKED ISSUE #1045: [RTL5] Incomplete spec implementation for channel DETACH/ATTACH
- Currently, as a part of [RTL5](https://sdk.ably.com/builds/ably/specification/main/features/#RTL5), [we only check for initialized, attaching, detaching and detached states](https://github.com/ably/ably-java/blob/dc4d464ed73084e1276eeccb20ba994472a112f7/lib/src/main/java/io/ably/lib/realtime/ChannelBase.java#L314).
- We are missing cases for the channel `suspended` and `failed` states, [RTL5b](https://sdk.ably.com/builds/ably/specification/main/features/#RTL5b) and [RTL5j](https://sdk.ably.com/builds/ably/specification/main/features/#RTL5j). 
- We have complete implementation for the same in case of [ably-go detach.](https://github.com/ably/ably-go/blob/a81632dfeb26bd545232674622463e5709d42e03/ably/realtime_channel.go#L429)
- These checks are needed as a part of [chat-kotlin-room-lifecycle internal rollbacks](https://github.com/ably-labs/ably-chat-kotlin/issues/6).
- Ideally, should also check for inconsistent connection states, CLOSING or FAILED as per [RTL5g](https://sdk.ably.com/builds/ably/specification/main/features/#RTL5g)
- Need to add unit tests for the same.

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-5117) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by sacOO7**
Fixed #1045 

## Summary by CodeRabbit

- **New Features**
  - Enhanced channel state management during attachment and detachment operations.
  - Improved handling for channel states such as `failed` and `suspended`.

- **Bug Fixes**
  - Resolved issues related to publishing messages when the channel is in a `failed` or `suspended` state.

- **Tests**
  - Added new test cases to validate channel behavior in various states, ensuring robust testing of attachment and detachment functionalities.

### CODE REVIEW

**Review by ttypic — APPROVED** [review]
LGTM

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimeChannelTest.java [hunk 1]
```diff
@@ -944,6 +944,46 @@ public void attach_success_callback() {
...
+            ably = new AblyRealtime(opts);
+
+            /* wait until connected */
+            (new ConnectionWaiter(ably.connection)).waitFor(ConnectionState.connected);
+            assertEquals("Verify connected state reached", ConnectionState.connected, ably.connection.state);
+
+            /* create a channel and attach */
+            final Channel channel = ably.channels.get("attach_success");
+            channel.state = ChannelState.failed;
+            channel.reason = new ErrorInfo("channel in failed state", 40000);
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeChannelTest.java:961]: We also need to make sure connection goes into `failed` state, can you suggest code for the same

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimeChannelTest.java [hunk 2]
```diff
@@ -1015,6 +1055,67 @@
...
+    public void detach_success_callback_on_suspended_state() {
+        AblyRealtime ably = null;
+        try {
+            ClientOptions opts = new ClientOptions("dummy_key");
+            opts.autoConnect = false;
+            ably = new AblyRealtime(opts);
+
+            /* create a channel and attach */
+            final Channel channel = ably.channels.get("detach_success");
+            channel.state = ChannelState.suspended;
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeChannelTest.java]: Fixed https://github.com/ably/ably-java/pull/1046/commits/47cd42fb93c11afb22b1b6804db14253b5b95ff7

**Comment by sacOO7** [comment]
@coderabbitai resolve

**Pull request merged**